### PR TITLE
CHAT-825 : remove user token in case of logout only and make sure token is always initialized

### PR DESCRIPTION
**Pull request by thomasdelhomenie**
Token was previously removed as soon as the server session has expired, which can happen even if no explicit logout has been requested.
When this event occurred, the Session Timeout message is displayed, and when clicking on Login the page is refreshed but an error is displayed. Since the Chat application has already been initialized, it is not initialized again and the token is not generated again for the user, so all requests return a 403.

This fix remove the user token only in case of logout and make sure the token is correctly initialized each time the application is loaded.

### CODE REVIEW

**Comment by thomasdelhomenie** [comment]
> This will lead to performance regression (This method is ued by mini chat icon on toolbar).

No, it is used only in full mode. For the mini chat on top bar the Notification application is used (org.exoplatform.chat.portlet.notification.NotificationApplication).

> IMO we should manage this through ListenerService by renewing the token on ChatApplication when the user login again.

We need to send the token invalidation event on logout, so we cannot rely only on login.

**Comment by boubaker** [comment]
> No, it is used only in full mode. For the mini chat on top bar the Notification application is used (org.exoplatform.chat.portlet.notification.NotificationApplication).

ok, will we have a different behavior in mini chat and standalone application ?

> We need to send the token invalidation event on logout, so we cannot rely only on login.

I meant the token renewal. But anyway, when the index will be called, a new instance will be added anyway and we will have profileInitialized_ set to false anyway, am I right ? I yes, I think no need to move token generation and user update each time we refresh the Chat application, WDYT ?

**Comment by boubaker** [comment]
As discussed we will try to improve the token invalidation on logout time, to check if there are still open sessions for user or not. If no other session, delete the token, else, keep it since it's used by other browsers sessions for the same user

File: application/src/main/java/org/exoplatform/chat/portlet/chat/ChatApplication.java [hunk 1]
```diff
@@ -207,17 +207,17 @@ public ChatApplication(OrganizationService organizationService, SpaceService spa
    */
   public void initChatProfile() {
 
+    // Generate token
+    token_ = ServerBootstrap.getToken(remoteUser_);
+
+    // Add User in the DB if it does not exist, or update it if it exists with a different token
+    ServerBootstrap.addUser(remoteUser_, token_, dbName);
```
Comment (boubaker) [inline application/src/main/java/org/exoplatform/chat/portlet/chat/ChatApplication.java]: IMO we should execute this only if retrieved token by `ServerBootstrap.getToken` is different from current used token, WDYT ?

**Comment by boubaker** [comment]
As discussed, the approach is very good, so that the logout is effective on chat only on the dedicated browser (does not affect other sessions of the same user).
But, we still be able to send and receive messages from logged out browser session, so to make sure that the websocket channel is cut on logout, we should modify method ChatApplication.prototype.sessionTimeout to cut the websocket connection from client side (at least) and from server side if possible when the logout even is triggered.

**Pull request merged**